File: apps/mobile/ios/HealthCoachAI/Services/AuthService.swift
import Foundation
import Combine

class AuthService: ObservableObject {
    @Published var isAuthenticated = false
    @Published var currentUser: User?
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let baseURL = ProcessInfo.processInfo.environment["API_BASE_URL"] ?? "http://localhost:8080"
    private let keychain = KeychainService()
    private var cancellables = Set<AnyCancellable>()
    
    // MARK: - Authentication State Management
    
    init() {
        loadStoredCredentials()
    }
    
    // MARK: - Email/Password Authentication
    
    func login(email: String, password: String) {
        isLoading = true
        errorMessage = nil
        
        let loginRequest = LoginRequest(email: email, password: password)
        
        guard let url = URL(string: "\(baseURL)/auth/login/email") else {
            setError("Invalid URL")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try JSONEncoder().encode(loginRequest)
        } catch {
            setError("Failed to encode request")
            return
        }
        
        URLSession.shared.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: AuthResponse.self, decoder: JSONDecoder())
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.setError("Login failed: \(error.localizedDescription)")
                    }
                },
                receiveValue: { [weak self] response in
                    self?.handleAuthSuccess(response)
                }
            )
            .store(in: &cancellables)
    }
    
    func register(email: String, password: String, firstName: String?, lastName: String?) {
        isLoading = true
        errorMessage = nil
        
        let registerRequest = RegisterRequest(
            email: email,
            password: password,
            firstName: firstName,
            lastName: lastName
        )
        
        guard let url = URL(string: "\(baseURL)/auth/register/email") else {
            setError("Invalid URL")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try JSONEncoder().encode(registerRequest)
        } catch {
            setError("Failed to encode request")
            return
        }
        
        URLSession.shared.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: AuthResponse.self, decoder: JSONDecoder())
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.setError("Registration failed: \(error.localizedDescription)")
                    }
                },
                receiveValue: { [weak self] response in
                    self?.handleAuthSuccess(response)
                }
            )
            .store(in: &cancellables)
    }
    
    // MARK: - Phone/OTP Authentication
    
    func sendOTP(phoneNumber: String, completion: @escaping (Bool) -> Void) {
        isLoading = true
        errorMessage = nil
        
        let otpRequest = OTPRequest(phoneNumber: phoneNumber)
        
        guard let url = URL(string: "\(baseURL)/auth/otp/send") else {
            setError("Invalid URL")
            completion(false)
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try JSONEncoder().encode(otpRequest)
        } catch {
            setError("Failed to encode request")
            completion(false)
            return
        }
        
        URLSession.shared.dataTaskPublisher(for: request)
            .map(\.data)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] result in
                    self?.isLoading = false
                    if case .failure(let error) = result {
                        self?.setError("Failed to send OTP: \(error.localizedDescription)")
                        completion(false)
                    }
                },
                receiveValue: { _ in
                    completion(true)
                }
            )
            .store(in: &cancellables)
    }
    
    func verifyOTP(phoneNumber: String, otp: String) {
        isLoading = true
        errorMessage = nil
        
        let verifyRequest = OTPVerifyRequest(phoneNumber: phoneNumber, otp: otp)
        
        guard let url = URL(string: "\(baseURL)/auth/otp/verify") else {
            setError("Invalid URL")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try JSONEncoder().encode(verifyRequest)
        } catch {
            setError("Failed to encode request")
            return
        }
        
        URLSession.shared.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: AuthResponse.self, decoder: JSONDecoder())
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isLoading = false
                    if case .failure(let error) = completion {
                        self?.setError("OTP verification failed: \(error.localizedDescription)")
                    }
                },
                receiveValue: { [weak self] response in
                    self?.handleAuthSuccess(response)
                }
            )
            .store(in: &cancellables)
    }
    
    // MARK: - Token Management
    
    func refreshToken() {
        guard let refreshToken = keychain.getRefreshToken() else {
            logout()
            return
        }
        
        let refreshRequest = RefreshTokenRequest(refreshToken: refreshToken)
        
        guard let url = URL(string: "\(baseURL)/auth/refresh") else {
            logout()
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        do {
            request.httpBody = try JSONEncoder().encode(refreshRequest)
        } catch {
            logout()
            return
        }
        
        URLSession.shared.dataTaskPublisher(for: request)
            .map(\.data)
            .decode(type: AuthResponse.self, decoder: JSONDecoder())
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure = completion {
                        self?.logout()
                    }
                },
                receiveValue: { [weak self] response in
                    self?.handleAuthSuccess(response)
                }
            )
            .store(in: &cancellables)
    }
    
    // MARK: - Logout
    
    func logout() {
        DispatchQueue.main.async {
            self.isAuthenticated = false
            self.currentUser = nil
            self.errorMessage = nil
            self.keychain.clearTokens()
        }
    }
    
    // MARK: - Private Methods
    
    private func handleAuthSuccess(_ response: AuthResponse) {
        currentUser = response.user
        isAuthenticated = true
        
        // Store tokens securely
        keychain.storeAccessToken(response.accessToken)
        keychain.storeRefreshToken(response.refreshToken)
        
        // Store user data
        if let userData = try? JSONEncoder().encode(response.user) {
            UserDefaults.standard.set(userData, forKey: "currentUser")
        }
    }
    
    private func loadStoredCredentials() {
        // Check if we have stored tokens
        guard keychain.getAccessToken() != nil,
              let userData = UserDefaults.standard.data(forKey: "currentUser"),
              let user = try? JSONDecoder().decode(User.self, from: userData) else {
            return
        }
        
        currentUser = user
        isAuthenticated = true
        
        // Refresh token to ensure it's still valid
        refreshToken()
    }
    
    private func setError(_ message: String) {
        DispatchQueue.main.async {
            self.errorMessage = message
            self.isLoading = false
        }
    }
}

// MARK: - Keychain Service

class KeychainService {
    private let service = "com.healthcoachai.app"
    
    func storeAccessToken(_ token: String) {
        store(token, forKey: "accessToken")
    }
    
    func storeRefreshToken(_ token: String) {
        store(token, forKey: "refreshToken")
    }
    
    func getAccessToken() -> String? {
        return retrieve(forKey: "accessToken")
    }
    
    func getRefreshToken() -> String? {
        return retrieve(forKey: "refreshToken")
    }
    
    func clearTokens() {
        delete(forKey: "accessToken")
        delete(forKey: "refreshToken")
    }
    
    private func store(_ value: String, forKey key: String) {
        let data = value.data(using: .utf8)!
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecValueData as String: data
        ]
        
        SecItemDelete(query as CFDictionary)
        SecItemAdd(query as CFDictionary, nil)
    }
    
    private func retrieve(forKey key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        guard status == errSecSuccess,
              let data = result as? Data,
              let value = String(data: data, encoding: .utf8) else {
            return nil
        }
        
        return value
    }
    
    private func delete(forKey key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]
        
        SecItemDelete(query as CFDictionary)
    }
}

File: apps/mobile/android/app/src/main/java/com/healthcoachai/app/data/services/AuthService.kt
package com.healthcoachai.app.data.services

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKeys
import com.google.gson.Gson
import com.healthcoachai.app.data.models.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*

interface AuthApi {
    @POST("auth/login/email")
    suspend fun login(@Body request: LoginRequest): Response<AuthResponse>
    
    @POST("auth/register/email")
    suspend fun register(@Body request: RegisterRequest): Response<AuthResponse>
    
    @POST("auth/otp/send")
    suspend fun sendOTP(@Body request: OTPRequest): Response<Map<String, Any>>
    
    @POST("auth/otp/verify")
    suspend fun verifyOTP(@Body request: OTPVerifyRequest): Response<AuthResponse>
    
    @POST("auth/refresh")
    suspend fun refreshToken(@Body request: RefreshTokenRequest): Response<AuthResponse>
    
    @POST("auth/logout")
    suspend fun logout(@Header("Authorization") token: String): Response<Map<String, String>>
}

class AuthService(private val context: Context) {
    private val baseUrl = "http://10.0.2.2:8080/" // Android emulator localhost
    
    private val retrofit = Retrofit.Builder()
        .baseUrl(baseUrl)
        .addConverterFactory(GsonConverterFactory.create())
        .build()
    
    private val api = retrofit.create(AuthApi::class.java)
    private val gson = Gson()
    
    private val _authState = MutableStateFlow(AuthState())
    val authState: StateFlow<AuthState> = _authState
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage
    
    private val encryptedPrefs: SharedPreferences by lazy {
        val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
        EncryptedSharedPreferences.create(
            "auth_prefs",
            masterKeyAlias,
            context,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }
    
    init {
        loadStoredCredentials()
    }
    
    suspend fun login(email: String, password: String): Result<AuthResponse> {
        return try {
            _isLoading.value = true
            _errorMessage.value = null
            
            val response = api.login(LoginRequest(email, password))
            
            if (response.isSuccessful) {
                val authResponse = response.body()!!
                handleAuthSuccess(authResponse)
                Result.success(authResponse)
            } else {
                val error = "Login failed: ${response.message()}"
                _errorMessage.value = error
                Result.failure(ApiError(error, response.code()))
            }
        } catch (e: Exception) {
            val error = "Login failed: ${e.message}"
            _errorMessage.value = error
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    suspend fun register(
        email: String,
        password: String,
        firstName: String?,
        lastName: String?
    ): Result<AuthResponse> {
        return try {
            _isLoading.value = true
            _errorMessage.value = null
            
            val response = api.register(RegisterRequest(email, password, firstName, lastName))
            
            if (response.isSuccessful) {
                val authResponse = response.body()!!
                handleAuthSuccess(authResponse)
                Result.success(authResponse)
            } else {
                val error = "Registration failed: ${response.message()}"
                _errorMessage.value = error
                Result.failure(ApiError(error, response.code()))
            }
        } catch (e: Exception) {
            val error = "Registration failed: ${e.message}"
            _errorMessage.value = error
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    suspend fun sendOTP(phoneNumber: String): Result<Boolean> {
        return try {
            _isLoading.value = true
            _errorMessage.value = null
            
            val response = api.sendOTP(OTPRequest(phoneNumber))
            
            if (response.isSuccessful) {
                Result.success(true)
            } else {
                val error = "Failed to send OTP: ${response.message()}"
                _errorMessage.value = error
                Result.failure(ApiError(error, response.code()))
            }
        } catch (e: Exception) {
            val error = "Failed to send OTP: ${e.message}"
            _errorMessage.value = error
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    suspend fun verifyOTP(phoneNumber: String, otp: String): Result<AuthResponse> {
        return try {
            _isLoading.value = true
            _errorMessage.value = null
            
            val response = api.verifyOTP(OTPVerifyRequest(phoneNumber, otp))
            
            if (response.isSuccessful) {
                val authResponse = response.body()!!
                handleAuthSuccess(authResponse)
                Result.success(authResponse)
            } else {
                val error = "OTP verification failed: ${response.message()}"
                _errorMessage.value = error
                Result.failure(ApiError(error, response.code()))
            }
        } catch (e: Exception) {
            val error = "OTP verification failed: ${e.message}"
            _errorMessage.value = error
            Result.failure(e)
        } finally {
            _isLoading.value = false
        }
    }
    
    suspend fun refreshToken(): Result<AuthResponse> {
        val refreshToken = getRefreshToken()
        
        return if (refreshToken != null) {
            try {
                val response = api.refreshToken(RefreshTokenRequest(refreshToken))
                
                if (response.isSuccessful) {
                    val authResponse = response.body()!!
                    handleAuthSuccess(authResponse)
                    Result.success(authResponse)
                } else {
                    logout()
                    Result.failure(ApiError("Token refresh failed", response.code()))
                }
            } catch (e: Exception) {
                logout()
                Result.failure(e)
            }
        } else {
            logout()
            Result.failure(Exception("No refresh token available"))
        }
    }
    
    fun logout() {
        _authState.value = AuthState()
        clearTokens()
    }
    
    private fun handleAuthSuccess(response: AuthResponse) {
        _authState.value = AuthState(
            isAuthenticated = true,
            user = response.user
        )
        
        // Store tokens securely
        storeAccessToken(response.accessToken)
        storeRefreshToken(response.refreshToken)
        
        // Store user data
        encryptedPrefs.edit()
            .putString("user_data", gson.toJson(response.user))
            .apply()
    }
    
    private fun loadStoredCredentials() {
        val accessToken = getAccessToken()
        val userDataJson = encryptedPrefs.getString("user_data", null)
        
        if (accessToken != null && userDataJson != null) {
            try {
                val user = gson.fromJson(userDataJson, User::class.java)
                _authState.value = AuthState(
                    isAuthenticated = true,
                    user = user
                )
                
                // Refresh token to ensure it's still valid
                // This would be done in a coroutine scope in real implementation
            } catch (e: Exception) {
                // Invalid stored data, clear everything
                logout()
            }
        }
    }
    
    private fun storeAccessToken(token: String) {
        encryptedPrefs.edit()
            .putString("access_token", token)
            .apply()
    }
    
    private fun storeRefreshToken(token: String) {
        encryptedPrefs.edit()
            .putString("refresh_token", token)
            .apply()
    }
    
    private fun getAccessToken(): String? {
        return encryptedPrefs.getString("access_token", null)
    }
    
    private fun getRefreshToken(): String? {
        return encryptedPrefs.getString("refresh_token", null)
    }
    
    private fun clearTokens() {
        encryptedPrefs.edit()
            .remove("access_token")
            .remove("refresh_token")
            .remove("user_data")
            .apply()
    }
}

data class AuthState(
    val isAuthenticated: Boolean = false,
    val user: User? = null
)